frontend/app/routes/_public+/intake+/$id+/communication-preference.tsx

```diff
@@ -26,46 +28,60 @@ export const handle = {
   pageTitleI18nKey: 'communication-preference:page-title',
 } as const satisfies RouteHandleData;
 
-export async function loader({ request }: LoaderFunctionArgs) {
+export async function loader({ request, params }: LoaderFunctionArgs) {
   const { COMMUNICATION_METHOD_DIGITAL_ID } = getClientEnv();
   const preferredLanguages = await getLookupService().getAllPreferredLanguages();
   const preferredCommunicationMethods = await getLookupService().getAllPreferredCommunicationMethods();
 
-  return json({ COMMUNICATION_METHOD_DIGITAL_ID, preferredLanguages, preferredCommunicationMethods });
+  const intakeFlow = getIntakeFlow();
+  const { state } = await intakeFlow.loadState({ request, params });
+
+  return json({ COMMUNICATION_METHOD_DIGITAL_ID, preferredLanguages, preferredCommunicationMethods, state: state.communicationPreferences });
 }
 
-export async function action({ request }: ActionFunctionArgs) {
-  const formDataSchema = z.object({
-    preferredLanguage: z.enum(['en', 'fr']),
-  });
+export async function action({ request, params }: ActionFunctionArgs) {
+  const intakeFlow = getIntakeFlow();
+  const { id } = await intakeFlow.loadState({ request, params });
+
+  const { COMMUNICATION_METHOD_DIGITAL_ID } = getClientEnv();
+
+  const formSchema = z
+    .object({
+      preferredMethod: z.string({ required_error: 'empty-radio' }),
+      email: z.string().min(1, { message: 'empty-field' }).email({ message: 'invalid-email' }).optional(),
+      confirmEmail: z.string().min(1, { message: 'empty-field' }).optional(),
+      preferredLanguage: z.string({ required_error: 'empty-radio' }),
+    })
+    .refine((allValues) => allValues.preferredMethod != COMMUNICATION_METHOD_DIGITAL_ID || (allValues.preferredMethod === COMMUNICATION_METHOD_DIGITAL_ID && allValues.email?.trim() === allValues.confirmEmail?.trim()), { message: 'not-the-same' });
```

Comment:
I find the `.refine()` call quite hard to read. What do you think about pulling it out into a separate function?

```suggestion
  const emailsMatch = (val: { preferredMethod: string; email?: string; confirmEmail?: string }) => {
    if (val.preferredMethod === COMMUNICATION_METHOD_DIGITAL_ID) {
      // emails have to match only when the preferred method is digital id
      return val.email?.trim() === val.confirmEmail?.trim();
    }

    return true;
  };

  const formSchema = z
    .object({
      preferredMethod: z.string({ required_error: 'empty-radio' }),
      email: z.string().min(1, { message: 'empty-field' }).email({ message: 'invalid-email' }).optional(),
      confirmEmail: z.string().min(1, { message: 'empty-field' }).optional(),
      preferredLanguage: z.string({ required_error: 'empty-radio' }),
    })
    .refine(emailsMatch, { message: 'not-the-same' });
```

Reply:
updated